cluster.go

```diff
@@ -289,6 +312,69 @@ func parseSlots(slots RedisMessage, defaultAddr string) map[string]group {
 	return groups
 }
 
+// parseShards - map redis shards for each redis nodes/addresses
+// defaultAddr is needed in case the node does not know its own IP
+func parseShards(shards RedisMessage, defaultAddr string, tls bool) map[string]group {
+	parseNodeEndpoint := func(msg map[string]RedisMessage) string {
+		endpoint := msg["endpoint"].string
+		switch endpoint {
+		case "":
+			return defaultAddr
+		case "?":
+			return ""
+		}
+
+		port := msg["port"].integer
+		if tls && msg["tls-port"].integer > 0 {
+			port = msg["tls-port"].integer
+		}
+
+		return net.JoinHostPort(endpoint, strconv.FormatInt(port, 10))
+	}
+
+	groups := make(map[string]group, len(shards.values))
+	for _, v := range shards.values {
+		slotsAndNodes, _ := v.ToMap()
+		var master string
+		nodes := slotsAndNodes["nodes"].values
+		for i := 0; i < len(nodes); i++ {
+			dict, _ := nodes[i].ToMap()
+			if dict["role"].string != "master" {
+				continue
+			}
+			master = parseNodeEndpoint(dict)
+			break
+		}
+
+		if master == "" {
+			continue
+		}
+
+		g, ok := groups[master]
+		if !ok {
+			g.slots = make([][2]int64, 0)
+			g.nodes = make([]string, 0, len(nodes)-1)
+			for i := 0; i < len(nodes); i++ {
+				dict, _ := nodes[i].ToMap()
+				dst := parseNodeEndpoint(dict)
+				if dst == "" {
+					continue
+				}
+				g.nodes = append(g.nodes, dst)
+			}
+		}
+		slots := slotsAndNodes["slots"]
+		arr, _ := slots.ToArray()
+		for i := 0; i < len(arr); i += 2 {
```

Comment:
Use `i + 1 < len(arr)` to make sure that the loop won't panic.
```suggestion
		for i := 0; i + 1 < len(arr); i += 2 {
```